Create replica

sploiselle opened the pull request:
Support for `CREATE CLUSTER REPLICA` with the following caveats:

- Does not yet support availability zones (keeping PR smaller); will add as fast follow once this gets approved
- Only handles name resolution per cluster, i.e. two clusters can have replicas of the same name. This also means `DROP CLUSTER REPLICA` requires specifying which cluster it's on (`DROP CLUSTER REPLICA bar FROM foo`). This was easier to implement rather than creating an additional type for tracking clusters and replica name collisions. If this is distasteful, glad to pivot and introduce another cluster/replica namespace.
- The proposed syntax for inline replicas isn't LL1 because replica options are also comma delimited. Instead replica options must be placed inside a paren (`CREATE CLUSTER a REPLICA b (option = a, option = b), REPLICA c (option = a, option = b)`), unless there is only one option in which case the paren can be omitted (`CREATE CLUSTER a REPLICA b option = a`).

### Motivation

This PR adds a known-desirable feature. #11778

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Let users create multiple replicas of a cluster to horizontally scale compute resources.

philip-stoev commented:
I would like to give this a spin tomorrow. That PR touches 37 files

philip-stoev commented:
Item No 1. I have a feeling that this is not strictly related to this PR, but I was unable to reproduce it with the SQL syntax that was available prior to it.

In main, creating a dataflow that panics on purposes causes computed to start crash-looping, as expected. Dropping the dataflow causes the crash loop to stop. With this branch, the crash loop continues even after the offending dataflow has been dropped. It seems that `SIZE '2-` is required to reproduce.

The panic is as follows:

```
2022-05-02T19:22:36.430702Z  WARN mz_dataflow_types::client::replicated: Rehydrating replica "cluster-default-r2"
thread 'thread 'timely:work-1timely:work-0' panicked at '' panicked at 'called `Option::unwrap()` on a `None` valuecalled `Option::unwrap()` on a `None` value', ', /home/pstoev/sploiselle-create-replica/src/compute/src/server.rs/home/pstoev/sploiselle-create-replica/src/compute/src/server.rs::212212::4141
... [ unreadable jumble of stack frames from multiple processes, all related to panicking and stack unwinding]
thread panicked while panicking. aborting.
``` 

To reproduce:

```
CREATE TABLE t1 (f1 TEXT);
CREATE MATERIALIZED VIEW v1 AS SELECT mz_internal.mz_panic(f1) FROM t1;
CREATE CLUSTER REPLICA r2 FOR default SIZE '2-2';
INSERT INTO t1 VALUES ('a');

SELECT mz_internal.mz_sleep(10);
DROP VIEW v1;
```

Please feel free to convert this into a separate ticket if this is out of scope of this PR.

benesch commented:
This looks really good, although that's relatively low confidence since this is about at the PR size where GitHub starts to melt. I'll have to take another pass at the fiddlier bits tomorrow.

> Only handles name resolution per cluster, i.e. two clusters can have replicas of the same name. This also means DROP CLUSTER REPLICA requires specifying which cluster it's on (DROP CLUSTER REPLICA bar FROM foo). This was easier to implement rather than creating an additional type for tracking clusters and replica name collisions. If this is distasteful, glad to pivot and introduce another cluster/replica namespace.

Nope, this was exactly the intent! Good catch. I sent #12213 to fix the design doc. Some more comments in that PR that I'd like your opinion on!

> The proposed syntax for inline replicas isn't LL1 because replica options are also comma delimited. Instead replica options must be placed inside a paren (CREATE CLUSTER a REPLICA b (option = a, option = b), REPLICA c (option = a, option = b)), unless there is only one option in which case the paren can be omitted (CREATE CLUSTER a REPLICA b option = a).

Oops! Also fixed in #12213. I think it'd be best to _always_ mandate the parens in the inline replica specification though—the extra complexity of the special case doesn't seem worth it to me!

philip-stoev commented:
Item No 2. `SHOW CLUSTER REPLICAS` requires experimental mode, but `CREATE CLUSTER REPLICA` does not require it.

philip-stoev commented:
Item No 3. What is "service name" ?

```
'cannot create multiple replicas with the same name 'size_1' or same service name 'cluster-foo-size-1' on cluster 'foo'
```

If this is a purely internal concept maybe it should not appear in error messages

philip-stoev commented:
Item No 4. `CREATE CLUSTER REPLICA IF NOT EXISTS` is accepted, but the `IF NOT EXISTS` part is ignored:

```
> CREATE CLUSTER REPLICA IF NOT EXISTS size_1 FOR foo SIZE '1';
rows didn't match; sleeping to see if dataflow catches up 50ms 75ms 112ms 169ms 253ms 335ms
^^^ +++
cluster-replicas.td:36:1: error: executing query failed: db error: ERROR: cannot create multiple replicas with the same name 'size_1' or same service name 'cluster-foo-size-1' on cluster 'foo': ERROR: cannot create multiple replicas with the same name 'size_1' or same service name 'cluster-foo-size-1' on cluster 'foo'
     |
  35 | 
  36 | > CREATE CLUSTER REPLICA IF NOT EXISTS size_1 FOR foo SIZE '1';
     | ^
```

philip-stoev commented:
Item No 5. Question, I guess also for @benesch 

If we support `option1=a, option2=b` syntax, why is `SIZE` an explicit keyword? Also, if we can have multiple options, why is the number of nodes and the number of workers per node both encoded in the `SIZE` argument? Should it be `CREATE CLUSTER REPLICA ... (machines=2, workers=2)` ?

philip-stoev commented:
Item No 6. Should testdrive also be dropping the default cluster and recreating it for every test?

This will increase the execution time but at the same time keep the tests isolated. Right now `cluster-replicas.td` manipulates the default cluster, and while it does return it back to the original state, it is on the test writer to make sure that this is always the case, which can lead to undesirable interactions between tests.

philip-stoev commented:
Item No 7. Small nit, there is somewhat of a disrepancy when handling unquoted `SIZE` identifiers.

```
! CREATE CLUSTER foo REPLICA size_2 SIZE 1;
contains: expected String or bare identifier

! CREATE CLUSTER foo REPLICA size_2 SIZE a;
contains: unknown cluster size
```

It may be best to mandate that `SIZE` is always quoted, to avoid any ambiguity?

philip-stoev reviewed (approved):
I took the liberty of pushing some more tests into the .td file in your branch . Please take that commit along for he ride to main -- squash it if you prefer.

jkosh44 commented on src/coord/src/coord.rs:
```diff
@@ -2668,22 +2636,74 @@ impl Coordinator {
+        if names.is_empty() {
+            return Ok(ExecuteResponse::DroppedComputeInstanceReplicas);
+        }
+        let cluster_name = cluster.unwrap();
```
This is subjective so feel free to ignore. It might make more sense for `DropComputeInstanceReplicaPlan` to be `Option<(Vec<String>, String)`, so you don't accidentally end up in a situation where `cluster` is `None`, but `names` is not empty.

sploiselle replied:
If names is not empty, cluster being none is a programming error and should panic, which is what the unwrap does--not quite sure what the difference is here?

jkosh44 replied:
Yeah sorry, I don't really think there is a practical difference. I was just thinking that `Option<(Vec<String>, String)` might make it harder for future programmers to make that error, though it's pretty subjective. I think it's fine the way it is.

jkosh44 reviewed (commented):
A couple of small things I noticed on my first pass.

jkosh44 commented on src/coord/src/catalog/storage.rs:
```diff
@@ -961,6 +1055,24 @@ struct DatabaseKey {
+struct ComputeInstanceReplicaValue {
+    #[prost(string)]
+    config: String,
+    #[prost(string)]
```
What I was trying to say is that for `compute_instance_replicas`, I don't think we need a a unique value check. Stash already has a unique key constraint, which in this case will check that same thing as the unique value check. Which also means that we probably don't need to persist `service_name` at all.

sploiselle replied:
Ah, this code is actually meaningfully exercised:

1. [`sequence_create_compute_instance_replica`](https://github.com/MaterializeInc/materialize/pull/12132/files#diff-9fc9aab48f6290b549b1d95ea71e6aa4225ac9fa870a8a0b8aba29108694291aR1970) generates the replica service name out of bounds, using the same function that determines the service name in the lower layers. This must be done here pre-catalog transaction because we cannot permit creating two replicas with the same service name.
- [`DataflowTypes::Client::Controller::add_replica_to_instance`](https://github.com/MaterializeInc/materialize/pull/12132/files#diff-9d19deda2eb11325e9505642a623ba52f9084383939ddef679283425a4fc8e57R189) uses this same function to generate the service name, which we've guaranteed will be unique during sequencing.

> However since service_name is made up of the compute name and the replica name, there's a 1-to-1 mapping of compute name to compute instance id, the ComputeInstanceReplicaKey is made up of compute instance id and replica name

Ah, I think you've misunderstood the code in `generate_replica_service_name`. For instance `"foo_replica"` and `"foo-replica"` are both valid replica identifiers but generate the same service name (unless we wanted to add a nonce to the service names names, which is legit, but not what the code does) because e.g. underscores are not valid in hostnames, so we just blithely replace them with dashes. We then have to ensure that we haven't generated multiple replicas of the same name, which is what all of the song and dance with the catalog checks does.

sploiselle replied:
I can add some comments to the code to express this relationship more clearly to future readers--sorry that this wasn't more obvious.

jkosh44 replied:
Ahh, ok that makes sense. Thanks for the explanation. In terms of the nonce I don't have a strong opinion, but I think it's probably not necessary for this PR.

jkosh44 replied:
@sploiselle Just had a random thought this morning. I don't think it's necessary for this PR, just more food for thought. If we added disallowed characters to replica names s.t. it matched all the changes done by `sanitize_str_for_service_hostname`, then I think we could probably remove `service_name` from here and remove the `sanitize_str_for_service_hostname`.

sploiselle commented on src/sql-parser/src/ast/defs/statement.rs:
```diff
@@ -940,6 +937,70 @@ impl<T: AstInfo> AstDisplay for ClusterOption<T> {
+    /// The replica's definition.
+    pub definition: ReplicaDefinition<T>,
+    /// Whether the `IF NOT EXISTS` clause was specified.
+    pub if_not_exists: bool,
```
Ah, good catch. Will remove. Ty!

mjibson commented on src/coord/src/catalog/storage.rs:
```diff
@@ -353,18 +367,20 @@ impl<S: Append> Connection<S> {
-                                .into(),
-                        )))
-                    }
+                let config: Option<ComputeInstanceIntrospectionConfig> = match v.config {
```
nit: `v.config.map(|config| ...)` might be more concise here.

mjibson replied:
Err nope, I must have misunderstood what was going on. Nevermind on my comment.

mjibson commented:
This will need a rebase from main to pick up the async catalog changes from https://github.com/MaterializeInc/materialize/pull/12208.

benesch reviewed (commented):
Shoot, sorry, I forgot to press send on this comment. It’s relevant to the discussion about replica name uniqueness.

sploiselle commented:
@philip-stoev

[Item No 1.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1115277083)

Yeah, the purview of this PR is simply tying the SQL/Coord code to the replica
API, so pathological cases are out of scope of this explicitly. However,
assessing replica health is still tbd on my end, so maybe there's something I'm
on the hook for still.

Either way, will file an issue on this.

[Item No 2.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1115981300)

Fixed

[Item No 3.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1115986944)

Fixed

[Item No 4.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1115988181)

Removed this entirely per [Joe's
suggestion](https://github.com/MaterializeInc/materialize/pull/12132#discussion_r863858258)

[Item No 5.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1115997094)

This would just be annoying to handle if the key were free-form in cases other
than `ReplicaOption::Remote`, which breaks the normal "with option" syntax. It's
certainly "fixable" (i.e. we can make it consistent with with options), but
doing so wouldn't be backward incompatible, so maybe we can punt until we know
we have more of the options pinned down?

[Item No 6.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1116001468)

What a good call! I think the better solution is to make the `default` cluster
immutable in testdrive so as to not make everyone pay the price. Note that all of the tests migrated to SLT anyway, but am leaving this in here.

[Item No 7.](https://github.com/MaterializeInc/materialize/pull/12132#issuecomment-1116004669)

This series of errors is maybe annoying, but the "correct" behavior for how
we've designed `WITH` options to work. Changing the behavior would make this one
option value differ from all the others, which I'm not convinced is complexity
worth introducing. If this proves onerous, glad to change behavior.

sploiselle commented on src/dataflow-types/src/client.rs:
```diff
@@ -56,17 +56,17 @@ pub type ComputeInstanceId = i64;
 #[derive(Clone, Debug, Serialize, Deserialize)]
-pub enum InstanceConfig {
-    /// Out-of-process named instance
+pub enum ConcreteComputeInstanceReplicaConfig {
```
This could get reduced down to `ConcreteReplicaConfig`; will throw that into the next commit that touches this code.

jkosh44 commented on src/dataflow-types/src/client.rs:
```diff
@@ -56,17 +56,17 @@ pub type ComputeInstanceId = i64;
+pub enum ConcreteComputeInstanceReplicaConfig {
+    /// Out-of-process replica
     Remote {
         /// A map from replica name to hostnames.
```
Bump, this is pretty minor but since the field is no longer a map, this comment may be a bit misleading.

sploiselle replied:
Oh derp--thanks for the ping on this.

antiguru commented on src/dataflow-types/src/client/replicated.rs:
```diff
@@ -81,6 +81,10 @@ where
         self.hydrate_replica(&identifier).await;
     }
 
+    pub fn get_replica_identifiers(&self) -> impl Iterator<Item = &String> {
```
Only used for an assertion in `controller.rs` -- is it necessary?

antiguru commented on src/dataflow-types/src/client/controller.rs:
```diff
@@ -131,6 +132,14 @@ impl Default for ClusterReplicaSizeMap {
 
+/// Deterministically generates replica names based on inputs.
+///
+/// This function needs to be publicly accessible so other layers can ensure we
```
Comment out of date -- function isn't `pub` anymore.

antiguru commented on src/dataflow-types/src/client/controller.rs:
```diff
@@ -229,20 +256,53 @@ where
+        config: ConcreteComputeInstanceReplicaConfig,
+    ) -> Result<(), anyhow::Error> {
+        let replica_name = generate_replica_service_name(instance_id, replica_id);
+        if let ConcreteComputeInstanceReplicaConfig::Managed {
+            size_config: _size_config,
+        } = config
+        {
+            let OrchestratorConfig { orchestrator, .. } = &self.orchestrator;
+            orchestrator
+                .namespace("compute")
+                .drop_service(&replica_name)
+                .await?;
+        }
+        let mut compute = self.compute_mut(instance_id).unwrap();
+        compute.remove_replica(&replica_name);
```
We might want this to be the other way around: first remove the replica, then tell the orchestrator about dropping the service. We might race for a moment otherwise when the service terminates and we try to reconnect. Not sure if it can actually happen due to being effectively single-threaded, but who knows.

antiguru commented on src/dataflow-types/src/client/controller.rs:
```diff
@@ -229,20 +256,53 @@ where
         instance: ComputeInstanceId,
     ) -> Result<(), anyhow::Error> {
         if let Some(mut compute) = self.compute.remove(&instance) {
+            assert!(
+                compute.client.get_replica_identifiers().next().is_none(),
+                "cannot drop instances with provisioned replicas; call `drop_replica` first"
+            );
```
I'm a bit torn about this `assert`, since we don't have many other assertions in `Controller`. Another take would be that dropping the instance automatically drop all replicas -- @frankmcsherry what's your take on this?

frankmcsherry replied:
It should probably return an `Err` variant, if we want the erroring behavior? I don't know if it should be CASCADE or not (I would say yes myself, because that is how most non-SQL things work). I think if this converts to an `Err` then it is fine to land and we can consider changing the behavior later.

sploiselle replied:
Just have this returning an error now

sploiselle commented:
Have @antiguru 's feedback implemented locally but will wait on word back on [this comment](https://github.com/MaterializeInc/materialize/pull/12132#discussion_r866637279) before pushing

philip-stoev commented:
Very curiously, trailing zeroes and underscores are ignored when converting replica names to service names. I think that is OK, was just a curious thing to see by accident:

```
materialize=> CREATE CLUSTER REPLICA "r1_C0" FOR default SIZE '1';
CREATE CLUSTER REPLICA
materialize=> CREATE CLUSTER REPLICA "r1_C00" FOR default SIZE '1';
ERROR:  cannot create multiple replicas with the same name 'r1_C00' or same service name 'cluster-default-r1-C' on cluster 'default'`
``

sploiselle commented:
@philip-stoev That error got removed entirely, but agreed it seems like something was amiss in that code.
